vendor/bat-native-ledger/src/bat/ledger/internal/legacy/client_properties.cc

```diff
@@ -47,4 +67,133 @@ bool ClientProperties::operator!=(
+  dict.Set(kWalletInfoKey, wallet_info.ToValue());
+  dict.Set(kBootTimestampKey, base::Int64ToValue(boot_timestamp));
+  dict.Set(kReconcileTimestampKey, base::Int64ToValue(reconcile_timestamp));
+  dict.Set(kFeeAmountKey, fee_amount);
+  dict.Set(kUserChangedFeeKey, user_changed_fee);
+  dict.Set(kRewardsEnabledKey, rewards_enabled);
+  dict.Set(kAutoContributeKey, auto_contribute);
+
+  base::Value::Dict inline_tips_dict;
+  for (const auto& [key, value] : inline_tips) {
+    inline_tips_dict.Set(key, value);
+  }
+  dict.Set(kInlineTipsKey, std::move(inline_tips_dict));
+
+  return dict;
+}
+
+bool ClientProperties::FromValue(const base::Value::Dict& dict) {
+  // Wallet Info
+  const auto* wallet_info_dict = dict.FindDict(kWalletInfoKey);
+  if (!wallet_info_dict) {
+    NOTREACHED();
+    return false;
+  }
+  if (!wallet_info.FromValue(*wallet_info_dict)) {
+    NOTREACHED();
+    return false;
+  }
+
+  // Boot Timestamp (There is no support for uint64_t. Writing JSON with such
+  // types violates the spec. As we need a uint64_t, we need to use an unsigned
+  // int and cast to a uint64_t)
+  if (auto value = base::ValueToInt64(dict.Find(kBootTimestampKey))) {
+    boot_timestamp = *value;
+  } else if (auto value = dict.FindDouble(kBootTimestampKey)) {
+    boot_timestamp = static_cast<uint64_t>(*value);
+  } else {
+    NOTREACHED();
+    return false;
+  }
+
+  // Reconcile Timestamp (There is no support for uint64_t. Writing JSON with
+  // such types violates the spec. As we need a uint64_t, we need to use an
+  // unsigned int and cast to a uint64_t)
+  if (auto value = base::ValueToInt64(dict.Find(kReconcileTimestampKey))) {
+    reconcile_timestamp = *value;
+  } else if (auto value = dict.FindDouble(kReconcileTimestampKey)) {
+    reconcile_timestamp = static_cast<uint64_t>(*value);
+  } else {
+    NOTREACHED();
+    return false;
+  }
+
+  // Fee Amount
+  if (auto value = dict.FindDouble(kFeeAmountKey)) {
+    fee_amount = *value;
+  } else {
+    NOTREACHED();
+    return false;
+  }
+
+  // User Changed Fee
+  if (auto value = dict.FindBool(kUserChangedFeeKey)) {
+    user_changed_fee = *value;
+  } else {
+    NOTREACHED();
+    return false;
+  }
+
+  // Auto Contribute
+  if (auto value = dict.FindBool(kAutoContributeKey)) {
+    auto_contribute = *value;
+  } else {
+    NOTREACHED();
+    return false;
+  }
+
+  // Rewards Enabled
+  if (auto value = dict.FindBool(kRewardsEnabledKey)) {
+    rewards_enabled = *value;
+  } else {
+    NOTREACHED();
+    return false;
+  }
+
+  // Inline Tips
+  if (const auto* value = dict.FindDict(kInlineTipsKey)) {
+    for (const auto inline_tip_value : *value) {
+      inline_tips.insert(
+          {inline_tip_value.first, inline_tip_value.second.GetBool()});
+    }
+  } else {
+    NOTREACHED();
+    return false;
+  }
+
+  return true;
+}
+
+std::string ClientProperties::ToJson() const {
+  std::string json;
+  base::JSONWriter::Write(ToValue(), &json);
```

Comment:
Not sure if `base::JSONWriter::Write()` leaves `json` intact on failure.
`return base::JSONWriter::Write(ToValue(), &json) ? json : {};`, or even the return type could be `absl::optional<std::string>`.

Reply:
I can change the code, but these functions shouldn't fail though, so returning empty is not ideal here.

Reply:
Done d608431c2ae7035267196a971607409479f6d869.